bigquery/google/cloud/bigquery/dataset.py

```diff
@@ -119,7 +124,7 @@ def project(self):
         :rtype: str
         :returns: the project (derived from the client).
         """
-        return self._client.project
+        return self._project or self._client.project
```

Comment:
Maybe so, given that a client instance normally won't change its project (leaving it dynamic allows for that possibility).